### DISCUSSION TITLE
VERSION from git describe ?

### DISCUSSION THREAD

**Discussion by jimc** [arighi/virtme-ng#41] [Ideas]
this patch gives me a legal version,

[jimc@frodo virtme-ng]$ vng -V
virtme-ng 1.17+v1.17-14-ga19857a

its arguably a bit noisy, but it could also be more deeply flawed
(whats it do when there is no repo, and could that even matter?)

so its an idea for now..

 
[jimc@frodo virtme-ng]$ git diff
diff --git a/virtme_ng/version.py b/virtme_ng/version.py
index fb930bc..d7a88a1 100644
--- a/virtme_ng/version.py
+++ b/virtme_ng/version.py
@@ -3,7 +3,10 @@
 
 """virtme-ng version"""
 
-VERSION = "1.17"
+import subprocess
+
+VERSION = "1.17+" + subprocess.check_output(
+        ['git', 'describe'], universal_newlines=True).strip()
 
 if __name__ == '__main__':
     print(VERSION)

**Comment by arighi** [comment]
I've been trying to figure out a reliable way to determine the version without manually updating `virtme-ng/version.py` every single time. I went to that route in the past (using `git describe`), but it makes the deb packaging a bit of a headache, because we don't have the git repo when the package is built... maybe we should add a condition, like if `.git/` exists then try to figure out the version via `git describe`, otherwise fallback to an hard-coded version... but I'm not yet 100% convinced this is a good idea...

**Reply by jimc** [reply]
I figured there might be packaging issues, and Im a python beginner.
My 1st try had a setver(), called from main. I think it tripped on setting an immutable 2x,
without a proper if then else, so I'll probly try again later.

.git/ exists seems a decent test, but brittle when PWD is different.
(Im often in ./builds/something, where git describe will work)
OTOH git describe might work oddly in dir-nested repos

MAN_VER="1.17"
AUTO_VER=`git describe`   # sorry perl-isms still come natural

.. then sort out the assignment to VERSION

**Comment by jimc** [comment]
<pre>
hows this look ?
its untested - will try soon.
- Im a python beginner, so will make some beginner mistakes for a while

# -*- mode: python -*-
# Copyright 2023 Andrea Righi

"""virtme-ng version"""

import subprocess

def get_version_string():
    try:
        # Run the `git describe` command and strip the leading 'v'
        version_output = subprocess.check_output(['git', 'describe'], universal_newlines=True).strip()
        return version_output[1:]
    except subprocess.CalledProcessError:
        # Git repository not found, return an empty string
        return ""

MAN_VERSION = "1.17"
GIT_VERSION = get_version_string()

# maintainer check for leading match on 2 versions
if GIT_VERSION.startswith(MAN_VERSION):
    version_string = GIT_VERSION[len(MAN_VERSION):].replace('-', '+')
    print("Using Git version:", version_string)
else:
    print("Git version mismatch:", GIT_VERSION)

VERSION = GIT_VERSION or MAN_VERSION

if __name__ == '__main__':
    print(VERSION)
<post>

**Reply by arighi** [reply]
Not bad at all! However, if `git` is not installed in the system (e.g., when building virtme-ng in an ubuntu ppa) your code would trigger a `FileNotFoundError` exception, unless we add git to the build-depends in the packaging info. And maybe in get_version_string() instead of returning an empty string if the git command fails you can return MAN_VERSION directly.

But even in this case MAN_VERSION needs to be updated and maintained, so I'd rather stick just with the manual version honestly...

Also I wouldn't add my name to the Copyright, since you wrote it feel free to add your name. :)

**Comment by jimc** [comment]
hi Andrea,  have you thought any more about versioning ?

python versioning, ala PEP440, disallows direct use of `git describe` output.

heres the error Im seeing atm, I think with the version I sent previously.

```
[jimc@frodo virtme-ng.git]$ make install
BUILD_VIRTME_NG_INIT=1 pip3 install --verbose -r requirements.txt .
Using pip 23.3.2 from /usr/lib/python3.12/site-packages/pip (python 3.12)
Defaulting to user installation because normal site-packages is not writeable
Processing /home/jimc/projects/virtme-ng.git
  Running command python setup.py egg_info
  Traceback (most recent call last):
    File "<string>", line 2, in <module>
    File "<pip-setuptools-caller>", line 34, in <module>
    File "/home/jimc/projects/virtme-ng.git/setup.py", line 154, in <module>
      setup(
    File "/usr/lib/python3.12/site-packages/setuptools/__init__.py", line 103, in setup
      return distutils.core.setup(**attrs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/lib/python3.12/site-packages/setuptools/_distutils/core.py", line 147, in setup
      _setup_distribution = dist = klass(attrs)
                                   ^^^^^^^^^^^^
    File "/usr/lib/python3.12/site-packages/setuptools/dist.py", line 314, in __init__
      self.metadata.version = self._normalize_version(self.metadata.version)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/lib/python3.12/site-packages/setuptools/dist.py", line 350, in _normalize_version
      normalized = str(Version(version))
                       ^^^^^^^^^^^^^^^^
    File "/usr/lib/python3.12/site-packages/setuptools/_vendor/packaging/version.py", line 198, in __init__
      raise InvalidVersion(f"Invalid version: '{version}'")
  setuptools.extern.packaging.version.InvalidVersion: Invalid version: '1.25-35-ga74e4d0'
  Using Git version: +35+ga74e4d0
  error: subprocess-exited-with-error
  
```

so that approach is out.

and `git describe` is prone to reading linux history when youre running vng there.

and google gemini tells me that PEP440 doesnt allow any extensions,
so me trying to replace + with - or # was futile

but setting a different accessory version isnt crazy.
esp if vng --version can display it as well as VERSION


is there a way to run git describe at pip install ?  
and perhaps punt and just say -dirty when in-tree ?

**Comment by arighi** [comment]
Hey @jimc what do you think about #131 ?

**Comment by marcosps** [comment]
@arighi given tht #131 was merged, this can closed then?

**Reply by arighi** [reply]
Yes, I think we can close this, thanks!
